src/QsCompiler/Transformations/CallGraph.cs

```diff
@@ -0,0 +1,569 @@
+                // This needs to be done first to cover an edge case.
+                foreach (var (typeParam, paramRes) in resolutionDictionary.Where(entry => mayBeReplaced.Contains(entry.Key)))
+                {
+                    // Get all the parameters whose value is dependent on the current resolution's type parameter,
+                    // and update their values with this resolution's value.
+                    foreach (var keyInCombined in mayBeReplaced[typeParam])
+                    {
+                        // Check that we are not doing an excessively constrictive resolution.
+                        success = success && !IsConstrictiveResolution(keyInCombined, paramRes);
+                        combinedBuilder[keyInCombined] = paramRes;
+                    }
+                }
+
+                // Add all resolutions to the current dictionary.
+                foreach (var (typeParam, paramRes) in resolutionDictionary)
+                {
+                    // Check that we are not doing an excessively constrictive resolution.
+                    success = success && !IsConstrictiveResolution(typeParam, paramRes);
+                    // Check that there is no conflicting resolution already defined.
+                    var conflictingResolutionExists = combinedBuilder.TryGetValue(typeParam, out var current)
+                        && !current.Equals(paramRes) && !IsSelfResolution(typeParam, current);
+                    success = success && !conflictingResolutionExists;
+                    combinedBuilder[typeParam] = paramRes;
+                }
+            }
+
+            combined = combinedBuilder.ToImmutable();
+            return success;
+        }
+    }
+
+    /// <summary>
+    /// Struct containing information that exists on edges in a call graph.
+    /// </summary>
+    public struct CallGraphEdge
+    {
+        public TypeParameterResolutions ParamResolutions;
+    }
+
+    /// <summary>
+    /// Struct containing information that exists on nodes in a call graph.
+    /// </summary>
+    public struct CallGraphNode
+    {
+        public QsQualifiedName CallableName;
+        public QsSpecializationKind Kind;
+        public QsNullable<ImmutableArray<ResolvedType>> TypeArgs;
+    }
+
+    /// <summary>
+    /// Comparer class used to check for equality between call graph edges.
+    /// Uses the Singleton pattern.
+    /// </summary>
+    public class CallGraphEdgeComparer : IEqualityComparer<CallGraphEdge>
+    {
+        public static CallGraphEdgeComparer Instance { get; } = new CallGraphEdgeComparer();
+
+        private CallGraphEdgeComparer() { }
+
+        public bool Equals(CallGraphEdge x, CallGraphEdge y) =>
+            x.ParamResolutions.ToImmutableHashSet().SetEquals(y.ParamResolutions.ToImmutableHashSet());
+
+        public int GetHashCode(CallGraphEdge obj) =>
+            obj.ParamResolutions.Aggregate(0, (hash, kvp) => hash ^ kvp.GetHashCode());
```

Comment:
Hmm, the bitwise or here doesn't seem particularly safe. 
Instead of defining and equality comparer, you could go with something like 
`private bool AreSameEdges(CallGraphEdge e1, CallGraphEdge e2) =>
     e1.ParamResolutions.OrderBy(kvp => kvp.Key).SequenceEqual(e2.ParamResolutions.OrderBy(kvp => kvp.Key));
`
and 
` else if (!edges.Any(e => AreSameEdges(e, edge)))`
in RecordDependency instead.

Reply:
The bitwise xor is a particularly well known way of combining hashes, and has the feature of ignoring order, which is desired here. Hashing is much faster than sorting, so it make sense to avoid using OrderBy in favor of this approach.

Reply:
Please don't claim something is the "well known way of" doing things without at least a reference. 
Well known or not, it is still problematic since a ^ b ^ a = b, a ^ a = 0, etc. 
Please don't close a conversation expressing concern without fully understanding and addressing that concern.

Reply:
Fully understanding here means to clarify what the reason was for the concern. My initial comment was pretty short and not detailing much of what the concern is, instead giving an alternative suggestion. When you would like to keep something in its current state, then it would be good to ask what exactly the reason was for the concern that wasn't specified in detail before coming to a conclusion.

Reply:
This logic was moved to the CallGraphEdge class. I added some comments there explaining why this method of hashing is well-suited to our needs in this situation.

Reply:
Here is where the logic was moved to: https://github.com/microsoft/qsharp-compiler/blob/sccarda/TypeResolution/src/QsCompiler/Transformations/CallGraph.cs#L159.

Reply:
I am still somewhat sceptic about the override for Equality and GetHashCode in general, opposed to using something like a helper that I suggested. By far the most likely cases are the case where there is only one or at most two type parameters, so just ordering them instead might be more efficient. I'll leave it up to @irinayat-MS to work with you to assess what makes sense.

Reply:
As far as I understand, the overrides are necessary to use CallGraphEdge objects as keys in dictionaries. We (Scott and I) reviewed the implementation of both methods and convinced ourselves that they are correct, that is, objects that are considered equal always return equal hashes. As far as quality of the hash function is concerned, I don't have any intuitions. Assuming high quality of C# GetHashCode for tuples and that there are likely to be only a handful of items in ParamResolutions dictionary, XOR should produce a reasonable hash as well...